src/controller/proxy/controller.go

```diff
@@ -0,0 +1,225 @@
+	ProxyManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo, w http.ResponseWriter) error
+}
+type controller struct {
+	blobCtl     blob.Controller
+	registryMgr registry.Manager
+	artifactCtl artifact.Controller
+	local       localInterface
+}
+
+// ControllerInstance -- Get the proxy controller instance
+func ControllerInstance() (Controller, error) {
+	// Lazy load the controller
+	// Because LocalHelper is not ready unless core startup completely
+	ctlLock.Lock()
+	defer ctlLock.Unlock()
+	if ctl == nil {
+		helper, err := newLocalHelper()
+		if err != nil {
+			return nil, err
+		}
+		ctl = &controller{
+			blobCtl:     blob.Ctl,
+			registryMgr: registry.NewDefaultManager(),
+			artifactCtl: artifact.Ctl,
+			local:       helper,
+		}
+	}
+	return ctl, nil
+}
+
+func (c *controller) UseLocal(ctx context.Context, art lib.ArtifactInfo) bool {
+	if len(art.Digest) > 0 {
+		exist, err := c.local.BlobExist(ctx, art.Digest)
+		if err == nil && exist {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *controller) ProxyManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo, w http.ResponseWriter) error {
+	var man distribution.Manifest
+	var err error
+	remoteRepo := art.ProxyCacheRemoteRepo()
+	r := newRemoteHelper(p.RegistryID)
+	ref := art.Digest
+	if len(ref) == 0 {
+		ref = art.Tag
+	}
+	man, err = r.Manifest(remoteRepo, ref)
+	if err != nil {
+		if errors.IsNotFoundErr(err) {
+			go func() {
+				c.local.DeleteManifest(remoteRepo, art.Tag)
+			}()
+		}
+		return err
+	}
+
+	ct, payload, err := man.Payload()
+	if err != nil {
+		return err
+	}
+	setHeaders(w, int64(len(payload)), ct, art.Digest)
+	_, err = w.Write(payload)
+	if err != nil {
+		return err
+	}
+
+	// Push manifest in background
+	go func() {
+		c.waitAndPushManifest(ctx, remoteRepo, man, art, ct, r)
+	}()
+
+	return nil
+}
+
+func (c *controller) ProxyBlob(ctx context.Context, p *models.Project, art lib.ArtifactInfo, w http.ResponseWriter) error {
+	remoteRepo := art.ProxyCacheRemoteRepo()
+	log.Debugf("The blob doesn't exist, proxy the request to the target server, url:%v", remoteRepo)
+	r := newRemoteHelper(p.RegistryID)
+	desc := distribution.Descriptor{}
+	size, bReader, err := r.BlobReader(remoteRepo, art.Digest)
+	if err != nil {
+		log.Errorf("failed to pull blob, error %v", err)
+		return err
+	}
+	defer bReader.Close()
+	// Use io.CopyN to avoid avoid out of memory when pulling big blob
+	written, err := io.CopyN(w, bReader, size)
+	if err != nil {
+		log.Errorf("failed to proxy the digest: %v, error %v", art.Digest, err)
+		return err
+	}
+	if written != size {
+		return errors.Errorf("The size mismatch, actual:%d, expected: %d", written, size)
+	}
+	desc.Size = size
+	desc.Digest = digest.Digest(art.Digest)
+
+	setHeaders(w, size, desc.MediaType, art.Digest)
+	// put blob to localHelper will start another connection to the remoteHelper,
+	// to reduce the impact of it, cache the blob after it send to the client
+	go func() {
```

Comment:
Put it before writing response to client?

Reply:
Add more comments here